catalog/core/catalog-core-api/src/main/java/ddf/catalog/security/SourceWarningsFilter.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.security;
+
+import ddf.catalog.operation.ProcessingDetails;
+import java.util.Set;
+
+/**
+ * <b> This code is experimental. While this interface is functional and tested, it may change or be
+ * removed in a future version of the library. </b>
+ */
+public interface SourceWarningsFilter {
+
+  /**
+   * Get the identifier of this {@code SourceWarningsFilter}
+   *
+   * @return either the identifier or, if this {@code SourceWarningsFilter} has no identifier, the
+   *     empty {@link String}
+   */
+  String getId();
+
+  /**
+   * Determine whether this {@code SourceWarningsFilter} can correctly filter out unreadable or
+   * useless {@code warnings} contained in the {@link ProcessingDetails} passed to it. If it can, so
+   * filter the {@link ProcessingDetails}'s {@code warnings} and create a {@link Set} composed
+   * exclusively of the {@code warnings} which pass through the filter.
```

Comment:
❓ Are we filtering or determining? What happens if this `SourceWarningsFilter` can't correctly filter? Does it throw an exception? Should we also have a `canFilter` function as well? Or is this documentation irrelevant because all `SourceWarningsFilter`s can actually filter any `ProcessingDetails` object it comes across?

Reply:
This ☝️ is probably the most crucial thing that needs resolution.

Reply:
I think we should change `getId()` to `canFilter()` and rewrite the Javadoc of this method to specify that no class should `filter()` until it has determined that it `canFilter()`. @mojogitoverhere ?

Reply:
Alternatively, we could specify in this Javadoc that `filter()` should return an empty `Set` if it determines that it can't correctly `filter()` the `ProcessingDetails` it receives as the argument.

Reply:
❗️ I wouldn't return an empty set if we can't correctly filter. Then how would we tell the difference between an inability to filter and a successful filter that yields empty?

Reply:
Using `canFilter()` would make the whole `SourceWarningsFilter` plugin system more generic. Right now, each `SourceWarningsFilter` is tied to a single source since we only use a filter if its `getId()` matches the source id of the `ProcessingDetails` object it is filtering. After talking with @kentmorrissey a while back, I think we agreed that using `canFilter()` would be more complex and we probably aren't going to need it (YAGNI) so we opted to stick with `getId()` for now.

To answer your questions @Lambeaux:
>Are we filtering or determining?

Filtering

>What happens if this `SourceWarningsFilter` can't correctly filter? Does it throw an exception?

I'm not sure I understand how the filtering could fail. The idea was that we would check if warning string matches some kind of pattern. If it matches, then keep the warning, otherwise filter it out.

>Should we also have a `canFilter` function as well? Or is this documentation irrelevant because all `SourceWarningsFilters` can actually filter any `ProcessingDetails` object it comes across?

Like I said above, we thought about using a `canFilter()` function but thought that added too much overhead. Instead each `SourceWarningsFilter` is tied directly to a source by its id.

Reply:
After talking in person we decided to go with the `canFilter()` approach. And to clarify the javadoc accordingly.